javascript/ql/src/semmle/javascript/frameworks/NodeJSLib.qll

```diff
@@ -506,4 +506,295 @@ module NodeJSLib {
+   */
+  abstract class RequestCallbackParam extends DataFlow::TrackedNode, RemoteFlowSource {
+  }
+  
+  /**
+   * A data flow node that is the parameter of a result callback for a Node.js HTTPS client request.
+   */
+  class HttpsRequestCallbackParam extends RequestCallbackParam {
+    HttpsRequest req;
+    
+    HttpsRequestCallbackParam() {
+      exists(DataFlow::MethodCallNode mcn, DataFlow::FunctionNode fn |
+        mcn = req and
+        fn.flowsTo(mcn.getArgument(1)) and
+        this = fn.getParameter(0)
+      )
+    }
+    
+    override string getSourceType() {
+      result = "https.request callback parameter"
+    }
+  }
+  
+  /**
+   * A data flow node that is  the parameter of a result callback for a Node.js HTTP client request.
+   */
+  class HttpRequestCallbackParam extends RequestCallbackParam {
+    HttpRequest req;
+    
+    HttpRequestCallbackParam() {
+      exists(DataFlow::MethodCallNode mcn, DataFlow::FunctionNode fn |
+        mcn = req and
+        fn.flowsTo(mcn.getArgument(1)) and
+        this = fn.getParameter(0)
+      )
+    }
+    
+    override string getSourceType() {
+      result = "http.request callback parameter"
+    }
+  }
+  
+  /**
+   * A data flow node that is the parameter of a data callback for an HTTP or HTTPS request made through Node.js.
+   */
+  class RequestCallbackData extends RemoteFlowSource {
+    RequestCallbackData() {
+      exists(RequestCallbackParam rcp, DataFlow::MethodCallNode mcn, DataFlow::FunctionNode fn |
+        rcp.flowsTo(mcn.getReceiver()) and
+        mcn.getArgument(0).mayHaveStringValue("data") and
+        fn.flowsTo(mcn.getArgument(1)) and
+        fn.getParameter(0) = this
+      )
+    }
+    
+    override string getSourceType() {
+      result = "http.request data parameter"
+    }
+  }
+  
+  /**
+   * A data flow node that is a response callback for an HTTP or HTTPS request made through Node.js.
+   */
+  class ClientRequestResponseHandler extends DataFlow::FunctionNode {
+    ClientRequestResponseHandler() {
+      exists(DataFlow::MethodCallNode mcn, ClientRequest cr |
+        cr.flowsTo(mcn.getReceiver()) and
+        mcn.getMethodName() = "on" and
+        mcn.getArgument(0).mayHaveStringValue("response") and
+        flowsTo(mcn.getArgument(1))
+      )
+    }
+  }
+  
+  /**
+   * A data flow node that is the parameter of a response callback for an HTTP or HTTPS request made through Node.js.
+   */
+  class ClientRequestResponseEvent extends RemoteFlowSource, DataFlow::TrackedNode {
+    ClientRequestResponseEvent() {
+      exists(ClientRequestResponseHandler handler |
+        this = handler.getParameter(0)
+      )
+    }
+    
+    override string getSourceType() {
+      result = "ClientRequest response event"
+    }
+  }
+  
+  /**
+   * A data flow node that is a data callback for an HTTP or HTTPS request made through Node.js.
```

Comment:
Here and elsewhere, could you please add a brief example code snippet to the qldoc to clarify what is being modelled?